### ISSUE TITLE
menuconfig doesn't aggregate choices with the same symbol name

### ISSUE THREAD

**Issue by pinkfluid** [ulfalizer/Kconfiglib#51]
Example Kconfig:

```
choice TEST
    prompt "test"

    config A
        bool "A"

    config B
        bool "B"
endchoice

choice TEST
    prompt "more symbols"
    config C
        bool "C"
endchoice
```
I'm not sure if this is intended behavior or a bug, but oldconfig and menuconfig show the above choices rather differently.

oldconfig aggregates the two choices into a single selection prompt:

```
$ ~/.local/bin/oldconfig
test (defined at Kconfig:1)
> 1. A (A)
  2. B (B)
  3. C (C)
```

While menuconfig shows two distinct selection groups:
```
(top menu)

    test (A)  --->
    more symbols (A)  --->
```

**Comment by ulfalizer** [comment]
It's intended behavior, though a bit confusing here.

When you define a symbol/choice in multiple locations, you're still creating just a single symbol/choice, which gets the union of all the properties/symbols from the different definition locations.

In the `oldconfig` case, the interface is letting you pick among all the symbols in the choice (`choice.syms`).

In the `menuconfig` case, you're always just walking the menu tree. There, there are two different menu nodes for the choice. The first menu node (with prompt "test") has the menu nodes for `A` and `B` as its children, and the second menu node (with prompt "more symbols") has the menu node for `C` as its child.

The only special logic for choie symbols in the menuconfig interface is (1) how they're displayed, and (2) that selecting a choice symbol automatically jumps back to the parent menu.

What's shown between `()` in the menuconfig interface is still the selection of the entire choice (`node.item.selection`, where `item` is the choice).

As a side note, people sometimes (seen it once in the kernel) do weird stuff like having a symbol within a choice `depend on` the symbol before it, which makes it so that the symbol with the `depends on` is no longer considered a choice symbol (completely undocumented).

When configuring such a choice, you can end up with weird stuff like this, which shows that it's just the menu structure you're seeing:

![skarmbild fran 2018-08-24 15-29-07](https://user-images.githubusercontent.com/154768/44587261-b48da880-a7b2-11e8-862b-3d5c4444677e.png)

It seems named choices with multiple prompts are broken in the C `menuconfig` (can't select the items from one of the definitions at all, and only the symbols that appear within the other definition), so no reference for behavior there. I don't think it would be too tricky to change it so that you can select all symbols at both locations. Not sure what makes the most sense there though.

**Comment by ulfalizer** [comment]
I suspect that having a named choice with symbols added from two locations might be broken in general in the C tools.

**Comment by ulfalizer** [comment]
Hmm... leaning towards showing all choice options in `menuconfig` too. Less confusing.

Think it would work to special-case choices so that all child menu nodes from all definition locations are shown when you enter the choice. I'll look into it a bit later.

**Comment by ulfalizer** [comment]
I made the change in the menuconfig: https://github.com/ulfalizer/Kconfiglib/commit/13884e934ec8eda928234c6506ae27f0334ec31b

Let me know if you spot any issues. I'll wait a bit before pushing out a release with it.

**Comment by pinkfluid** [comment]
The choice case now works. What's interesting is that you can have a sourced file include entries to choices in the top Kconfig.

```
choice TEST
    prompt "test"
    config A
        bool "CHOICE_A"
    config B
        bool "CHOICE_B"
endchoice

config DUMMY1
    bool "Dummy1"


osource Kconfig.include
```

choice TEST
    prompt "test"
    depends on 0

    config C
        bool "CHOICE_C"

endchoice

```

The "depends on 0" hides the second choice. What you get is this:
![2018-08-25-205501_206x80_scrot](https://user-images.githubusercontent.com/1593069/44621600-33fda380-a8a9-11e8-8762-a71a514510c4.png)
![2018-08-25-205518_174x80_scrot](https://user-images.githubusercontent.com/1593069/44621602-3829c100-a8a9-11e8-91f8-fa8e71a2fbfb.png)

**Comment by pinkfluid** [comment]
Menuconfig still shows separate options:

```
menuconfig TEST
    bool "test"
    config A
        depends on TEST
        bool "CHOICE_A"
    config B
        depends on TEST
        bool "CHOICE_B"

config DUMMY1
    bool "Dummy1"

menuconfig TEST
    bool "test"
    config C
        bool "CHOICE_C"
        depends on TEST
```
![2018-08-25-210124_124x81_scrot](https://user-images.githubusercontent.com/1593069/44621659-10872880-a8aa-11e8-84ff-a0b2b7d7d6fa.png)
![2018-08-25-210134_136x69_scrot](https://user-images.githubusercontent.com/1593069/44621660-10872880-a8aa-11e8-8d9f-d0cf4a66f69f.png)

The version is v10.2.1-5-gdbd9c59 (dbd9c59a4de57089de9f39d38e451ce8e6a869ba)

**Comment by ulfalizer** [comment]
The `menuconfig` case is expected. Internally, you're creating this menu structure:

```
["test"]->["CHOICE_A"]
   |           |
   |           v
   |      ["CHOICE_B"]
   v
["test"]->["CHOICE_C"]
```

The two menu nodes with prompt "test" both point to the symbol TEST, and it can be configured at either location. If you toggle one of them, you'll see that the other one changes too. You can even use a different prompt at each location.

Lots of things depend on Kconfig working this way, so it's unlikely to change. The choice fix is more of a hack that merges the menu trees at the two locations (because I agree that it makes things less confusing there). In general, the menu tree is "static" though.

`<soapbox>`
Some general advice from experience is to strive for simplicity and avoid multiple definition locations unless you have a super good use for them (e.g., if the symbol definitions absolutely have to be dynamic). Many projects go a bit crazy with them, creating Kconfig configurations that are hard to understand.

IMO, having one long symbol definition isn't obviously worse if you need to support multiple boards or the like. It centralizes information (making stuff easier to find) and makes the behavior simple to understand.

Note that you can put an `if` on the prompt if the symbol should only be configurable on some boards and have some fixed value on others.

The Linux kernel uses very few multiply-defined symbols, FWIW.
`</soapbox>`

**Comment by ulfalizer** [comment]
The misunderstanding might be that you expected the second definition to change the first one too. It's more of an `OR` operation than an `AND` operation though. All the properties from the different definition locations end up on the final symbol/choice, and the menu nodes (which hold the prompts) for the different definition locations point to that finalized symbol.

**Comment by pinkfluid** [comment]
Makes sense -- the only problem with the kconfig syntax is that documentation is somewhat lacking regarding these obscure situations. 

I've already solved the original problem I had with orsource and globs (adding to a global choice list from several sub-kconfigs).

**Comment by ulfalizer** [comment]
@pinkfluid 
Yeah, `kconfig-language.txt` definitely lacks some details. Planning to send a patch to update it.

The docstring at the top of `kconfiglib.py` covers a lot more Kconfig nitty-gritty.
